// typesense/src/Typesense/MultiSearch.ts
import ApiCall from './ApiCall'
import Configuration from './Configuration'
import RequestWithCache from './RequestWithCache'
import { DocumentSchema, SearchParams, SearchResponse } from './Documents'

const RESOURCEPATH = '/multi_search'

export interface MultiSearchRequestSchema<T> extends SearchParams<T> {
  collection?: string
}

export interface MultiSearchRequestsSchema<T> {
  searches: MultiSearchRequestSchema<T>[]
}

export interface MultiSearchResponse<T> {
  results: SearchResponse<T>[]
}

export default class MultiSearch<T extends DocumentSchema = {}> {
  private requestWithCache: RequestWithCache

  constructor(
    private apiCall: ApiCall,
    private configuration: Configuration,
    private useTextContentType: boolean = false
  ) {
    this.requestWithCache = new RequestWithCache()
  }

  perform(
    searchRequests: MultiSearchRequestsSchema<T>,
    commonParams: Partial<MultiSearchRequestSchema<T>> = {},
    {
      cacheSearchResultsForSeconds = this.configuration.cacheSearchResultsForSeconds
    }: { cacheSearchResultsForSeconds?: number } = {}
  ): Promise<MultiSearchResponse<T>> {
    let additionalHeaders = {}
    if (this.useTextContentType) {
      additionalHeaders['content-type'] = 'text/plain'
    }

    let additionalQueryParams = {}
    if (this.configuration.useServerSideSearchCache === true) {
      additionalQueryParams['usecache'] = true
    }
    const queryParams = Object.assign({}, commonParams, additionalQueryParams)

    return this.requestWithCache.perform(
      this.apiCall,
      this.apiCall.post,
      [RESOURCEPATH, searchRequests, queryParams, additionalHeaders],
      { cacheResponseForSeconds: cacheSearchResultsForSeconds }
    )
  }
}
